### PR TITLE
Record/Playback Emulator Inputs

### PR DESCRIPTION
**Pull request by dylwhich**
### Description

This adds functionality to the emulator for recording inputs to a file and playing them back later. It also supports some extra non-input actions when playing back a recording, such as enabling fuzzing, saving a screenshot, force-changing modes, and exiting the emulator.

Oh, and it adds a headless argument to have a hidden window, but I'm not sure if that works outside of Linux / for all rawdraw drivers.

### Test Instructions

Record inputs by running `swadge_emulator --record recording.csv` or just `swadge_emulator --record` (which should name the file `rec-xxxxxxx.csv`). When you're done, close the emulator and the file should now be full of inputs!

Play back that file with `swadge_emulator --playback recording.csv` and it should do exactly the same things as when recorded.

Edit the recording, and try adding these lines to the end:
```
14500000,SetMode,Touch Test
15000000,Screenshot,
15000000,Fuzz,
17500000,Screenshot,filename.bmp
20000000,Quit,
```
If you play back the recording with `swadge_emulator --playback recording.csv` again, it should repeat your inputs and then, after the original script ends, will wait until almost 15 seconds have passed and switch to the touch test mode, take a screenshot (`screenshot-xx.bmp`), begin fuzzing, then take another screenshot 2.5s later, and finally quit after waiting another 2.5s. You should see a `screenshot-xx.bmp` and `filename.bmp` in your working directory, and they should look correct.

Then, delete the screenshots and run the emulator again with `--headless` as well; it should do everything the same, but no window should appear. The screenshots should still be generated and look normal.

### Readiness Checklist
- [x] I have run `make format` to format the changes
- [x] I have compiled the firmware and the changes have no warnings
- [x] I have compiled the emulator and the changes have no warnings
- [x] I have run `make cppcheck` and checked that `cppcheck_result.txt` has no warnings for the changes
- [x] I have added doxygen comments to any code used by more than one Swadge mode. This includes `/*! \file` comments with Design Philosophy, Usage, and Example sections for new headers.
- [x] I have run `make docs` and checked that `doxy_warnings.txt` has no warnings for the new code

### CODE REVIEW

File: emulator/src/emu_main.c [hunk 1]
```diff
@@ -140,13 +136,23 @@ int main(int argc, char** argv)
...
+        calculatePaneMinimums(paneMins);
         int32_t sidePanesW      = paneMins[PANE_LEFT].min + paneMins[PANE_RIGHT].min;
         int32_t topBottomPanesH = paneMins[PANE_TOP].min + paneMins[PANE_BOTTOM].min;
+        int32_t winW            = (TFT_WIDTH)*2 + sidePanesW;
+        int32_t winH            = (TFT_HEIGHT)*2 + topBottomPanesH;
+
+        if (emulatorArgs.headless)
+        {
+            // If the window dimensions are negative, the window will still exist but not be displayed.
+            // TODO does this work on all platforms?
```
Comment (AEFeinstein) [inline emulator/src/emu_main.c:149]: I don't know

File: emulator/src/extensions/emu_args.c [hunk 2]
```diff
@@ -156,11 +164,14 @@ static const optDoc_t argDocs[] =
...
     { 0,  argFuzzTouch,   "y|n",   "Set whether touchpad inputs are fuzzed" },
     { 0,  argFuzzMotion,  "y|n",   "Set whether motion inputs are fuzzed" },
+    { 0,  argHeadless,    NULL,    "Runs the emulator without a window." },
     { 0,  argHideLeds,    NULL,    "Don't draw simulated LEDs next to the display" },
     {'l', argLock,        NULL,    "Lock the emulator in the start mode" },
     {'m', argMode,        "MODE",  "Start the emulator in the swadge mode MODE instead of the main menu"},
     { 0,  argModeSwitch,  "TIME",  "Enable or set the timer to switch modes automatically" },
     { 0,  argModeList,    NULL,    "Print out a list of all possible values for MODE" },
+    {'p', argPlayback,    "FILE",  "Play back recorded emulator inputs from a file" },
+    {'r', argRecord,      "FILE",  "Record emulator inputs to a file" },
```
Comment (AEFeinstein) [inline emulator/src/extensions/emu_args.c:174]: I don't love having two separate arrays that have to stay in sync, but this is still pretty nice.

File: emulator/src/extensions/emu_args.c [hunk 3]
```diff
@@ -89,6 +89,8 @@ emuArgs_t emulatorArgs = {
     .fuzzTouch   = false,
     .fuzzMotion  = false,
 
+    .headless = false,
```
Comment (AEFeinstein) [inline emulator/src/extensions/emu_args.c:92]: Did you mean to add `.playback` and `.record` here? Also `.recordFile` and `.replayFile`

File: emulator/src/extensions/replay/ext_replay.c [hunk 4]
```diff
@@ -0,0 +1,868 @@
...
+
+    replay.lastTouchR         = touchR;
+    replay.lastTouchPhi       = touchPhi;
+    replay.lastTouchIntensity = touchIntensity;
+    replay.lastAccelX         = accelX;
+    replay.lastAccelY         = accelY;
+    replay.lastAccelZ         = accelZ;
+
+    // Flush all the entries to the file so that we can close the file the proper way
+    // which is obviously to let the OS deal with it when the process exits
```
Comment (AEFeinstein) [inline emulator/src/extensions/replay/ext_replay.c:318]: 😲

File: emulator/src/extensions/replay/ext_replay.c [hunk 5]
```diff
@@ -0,0 +1,868 @@
...
+ * @brief Play back any recorded actions queued for the given frame
+ *
+ * @param frame
+ */
+static void replayPlaybackFrame(uint64_t frame)
+{
+    // Unless we've finished reading the file completely
+    if (!replay.readCompleted)
+    {
+        int64_t time           = esp_timer_get_time();
```
Comment (AEFeinstein) [inline emulator/src/extensions/replay/ext_replay.c:332]: This uses `CLOCK_MONOTONIC`, good.

File: emulator/src/extensions/replay/ext_replay.c [hunk 6]
```diff
@@ -0,0 +1,868 @@
...
+        }
+        else
+        {
+            // Big Endian
+            fputc(vals[size - i - 1], stream);
+        }
+    }
+}
+
+bool takeScreenshot(const char* name)
```
Comment (AEFeinstein) [inline emulator/src/extensions/replay/ext_replay.c:760]: I was expecting STB PNG, but this works too

**Review by AEFeinstein — APPROVED** [review]

File: emulator/src/extensions/emu_args.c (hunk 3 above)
Comment (AEFeinstein) [reply emulator/src/extensions/emu_args.c:92]: @dylwhich this is the only comment that matters

Comment (dylwhich) [reply emulator/src/extensions/emu_args.c:92]: Ah, I've been terrible about keeping these consistent, yeah.

Comment (AEFeinstein) [reply emulator/src/extensions/emu_args.c:92]: Wanna update before I merge or no?

**Pull request merged**